### PR TITLE
[FB6918951] Expand "Make Memberwise Initializer" Refactoring Action

### PR DESCRIPTION
**Pull request by CodaFi**
- Struct declarations are now eligible for the refactoring action
- Default argument expressions specified in source are printed into the generated initializer as well
- IUOs are printed with a default argument of `nil`

Note that we are explicitly missing the ability to print a 'nil' default argument for members of type T? because TypeReprs are not fully resolved by the time the refactoring engine runs and the initializer is appended to the AST by the typechecker.

Resolves applefeedback://FB6918951 and rdar://50093984

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke tets

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -2752,78 +2752,107 @@ bool RefactoringActionLocalizeString::performChange() {
...
-    EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " +
-                        memberTypeVector[i]);
-    
-    if (i != memberTypeVector.size() - 1) {
-      EditConsumer.accept(SM, targetLocation, ", ");
+  auto insertMember = [&SM](const MemberwiseParameter &memberData,
+                            llvm::raw_ostream &OS, bool wantsSeparator) {
+    OS << memberData.Name << ": " << memberData.MemberType.getString();
+    if (auto *expr = memberData.DefaultExpr) {
+      if (isa<NilLiteralExpr>(expr)) {
```
Comment (harlanhaskins) [inline lib/IDE/Refactoring.cpp]: This whole check feels like it should be factored out and maybe made into something that `ParamDecl::getDefaultValueStringRepresentation` could use, in general.

As it stands, this is duplicating the whole purpose behind `DefaultArgumentKind`.

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -2752,78 +2752,107 @@ bool RefactoringActionLocalizeString::performChange() {
...
-  
-  for (size_t i = 0, n = memberTypeVector.size(); i < n ; i++) {
-    EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " +
-                        memberTypeVector[i]);
-    
-    if (i != memberTypeVector.size() - 1) {
-      EditConsumer.accept(SM, targetLocation, ", ");
+  auto insertMember = [&SM](const MemberwiseParameter &memberData,
+                            llvm::raw_ostream &OS, bool wantsSeparator) {
+    OS << memberData.Name << ": " << memberData.MemberType.getString();
```
Comment (harlanhaskins) [inline lib/IDE/Refactoring.cpp]: `memberData.MemberType.print(OS)`?

File: lib/IDE/Refactoring.cpp [hunk 3]
```diff
@@ -2752,78 +2752,107 @@ bool RefactoringActionLocalizeString::performChange() {
   return false;
 }
 
+struct MemberwiseParameter {
```
Comment (harlanhaskins) [inline lib/IDE/Refactoring.cpp]: Comments would be nice, if you could specify the invariants (DefaultExpr night be null, Type must be valid, Name must be valid)

**Review by harlanhaskins — CHANGES_REQUESTED** [review]

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (harlanhaskins) [reply lib/IDE/Refactoring.cpp]: This can be a follow-up patch. I was going to tackle this very soon too.

File: lib/IDE/Refactoring.cpp (hunk 2 above)
Comment (CodaFi) [reply lib/IDE/Refactoring.cpp]: Paying the same costs for a syntactically clearer result this way.

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (CodaFi) [reply lib/IDE/Refactoring.cpp]: Feels weird to allocate and throw away an entire ParamDecl here, but I see what you mean about sequestering the logic up in the AST - I think that primitive on ParamDecl can be generalized to accept this data instead.  Or the header with the convenience for printing default arguments can actually be moved to the includes directory so I can consume it here.  Didn’t want to touch any of that for such a simple patch.

Comment (harlanhaskins) [reply lib/IDE/Refactoring.cpp]: Yeah, it should be factored out of ParamDecl. I don’t think it should block this patch.

**Comment by CodaFi** [comment]
@nkcsgexi ping.

**Review by nkcsgexi — APPROVED** [review]
LGTM!

**Review by harlanhaskins — APPROVED** [review]

**Comment by CodaFi** [comment]
Thanks y’all.  Hopefully this can be cleaned up a little more later from @harlanhaskins’ refactoring plans.

**Comment by CodaFi** [comment]
⛵️

**Comment by CodaFi** [comment]
rdar://50093984

**Pull request merged**